app/deliver_tx.go

```diff
@@ -0,0 +1,23 @@
+package app
+
+import (
+	"fmt"
+
+	"github.com/celestiaorg/celestia-app/x/upgrade"
+	abci "github.com/tendermint/tendermint/abci/types"
+)
+
+func (app *App) DeliverTx(req abci.RequestDeliverTx) abci.ResponseDeliverTx {
+	sdkTx, err := app.txConfig.TxDecoder()(req.Tx)
+	if err == nil {
+		if appVersion, ok := upgrade.IsUpgradeMsg(sdkTx.GetMsgs()); ok {
+			if !IsSupported(appVersion) {
+				panic(fmt.Sprintf("network has upgraded to version %d which is not supported by this node. Please upgrade and restart", appVersion))
+			}
+			app.SetProtocolVersion(appVersion)
```

Comment:
if this is meant to be done as the first transaction, should we just do this in BeginBlock?

Reply:
I think this would be more complex as we would need to track all the different proposed blocks and match them with the one that eventually gets committed.

Reply:
This will be easier in finalize block because it can reside as part of the `preBlock` call back function